src/lib/pubkey/classic_mceliece/cmce_decaps.h

```diff
@@ -0,0 +1,84 @@
+#include <botan/internal/cmce_field_ordering.h>
+#include <botan/internal/cmce_keys_internal.h>
+#include <botan/internal/cmce_matrix.h>
+#include <botan/internal/pk_ops.h>
+#include <botan/internal/pk_ops_impl.h>
+
+namespace Botan {
+
+/**
+ * Classic McEliece Decapsulation Operation
+ */
+class BOTAN_TEST_API Classic_McEliece_Decryptor final : public PK_Ops::KEM_Decryption_with_KDF {
+   public:
+      /**
+       * @brief Constructs a Classic_McEliece_Decryptor object with the given private key.
+       * @param key The private key used for decryption.
+       */
+      Classic_McEliece_Decryptor(std::shared_ptr<Classic_McEliece_PrivateKeyInternal> key, std::string_view kdf) :
+            KEM_Decryption_with_KDF(kdf), m_key(std::move(key)) {}
+
+      size_t raw_kem_shared_key_length() const override { return m_key->params().hash_out_bytes(); }
+
+      size_t encapsulated_key_length() const override { return m_key->params().ciphertext_size(); }
+
+      void raw_kem_decrypt(std::span<uint8_t> out_shared_key, std::span<const uint8_t> encapsulated_key) override;
+
+   private:
+      /**
+       * @brief Computes the syndrome of a code word.
+       *
+       * Corresponds to H' * code_word of the spec, where H' is the syndrome computation matrix used for the
+       * Berlekamp's method of decoding. See https://tungchou.github.io/papers/mcbits.pdf for more information.
+       *
+       * @param params The McEliece parameters.
+       * @param goppa_poly The Goppa polynomial.
+       * @param ordering The field ordering.
+       * @param code_word The code word.
+       * @return The syndrome S(x) of the code word.
+       */
+      Classic_McEliece_Polynomial compute_goppa_syndrome(const Classic_McEliece_Parameters& params,
+                                                         const Classic_McEliece_Minimal_Polynomial& goppa_poly,
+                                                         const Classic_McEliece_Field_Ordering& ordering,
+                                                         const secure_bitvector& code_word);
+
+      /**
+       * @brief Applies the Berlekamp-Massey algorithm to compute the error locator polynomial given a syndrome.
+       *
+       * The error locator polynomial C can be used for decoding, as C(a_i) = 0 <=> error at position i.
+       *
+       * @param params The McEliece parameters.
+       * @param syndrome The syndrome polynomial of the code word.
+       * @return The error locator polynomial.
+       */
+      Classic_McEliece_Polynomial berlekamp_massey(const Classic_McEliece_Parameters& params,
+                                                   const Classic_McEliece_Polynomial& syndrome);
+
+      /**
+       * @brief Decodes a code word using Berlekamp's method.
+       *
+       * @param big_c The code word.
+       * @return A pair containing the decoded message and the error pattern.
+       */
+      std::pair<CT::Mask<uint8_t>, secure_bitvector> decode(bitvector big_c);
```

Comment:
Given that this class doesn't have members other than the (hopefully read-only) `m_key`, these helpers could probably be marked as const.